[Woo POS] Improve Point of Sale image retrieval

joshheald opened the pull request:
<!-- Remember about a good descriptive title. -->

Part of: #13175 
<!-- Id number of the GitHub issue this PR addresses. -->

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

Point of Sale images are always shown through SwiftUI.

We don't need to use the `Image(uiImage:)` wrapper for images in the app bundle, we can use `Image(_ name:)` instead – there's no reason for the extra layer of UIKit code.

Additionally, we previously had defined the names in an extension on String. They're not properties of a general string, and it will give strange autocompletion results in other parts of the app to keep this behaviour.

Before adding new images in #13175, I'm updating this code a little to improve the above.

This PR updates all our existing POS image calls to use `Image(_ name:)`, and moves the name definitions to a `PointOfSaleAssets` enum. 

Note – a number of these images are SF Symbols exported as images, which is also questionable. We have #13458 to address that.

## Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->

1. Launch the app and select a Woo POS eligible store
2. Open POS mode
3. Connect a card reader
4. Run through a transaction, checking that the expected images all show correctly – see below for list.

## Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->

- [x] Exit POS icon in `...` menu
- [x] Get Support icon in `...` menu
- [x] X button to dismiss the simple products banner
- [x] Remove button for an item in the cart
- [x] Back button chevron in Checkout screen
- [x] Ready for Payment image when reader is done preparing on Checkout screen
- [x] Processing Payment image after tapping a card
- [x] Payment successful checkmark after payment
- [x] Curving back button on `New Transaction` button after successful payment

Card reader connection images are not affected, as they're the ones from the original app which will be replaced here.

## Screenshots
<!-- Include before and after images or gifs when appropriate. -->


https://github.com/user-attachments/assets/6fe9b326-260f-4c53-9d78-b863e47abb5b


---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.
- [ ] This PR includes refactoring; smoke testing of the entire section is needed.

joshheald committed:
13175 Remove unneccesary returns

joshheald committed:
13175 move image name definitions to enum

Previously these were in an extension on all Strings… which seemed strange as they’re only really relevant to loading POS images.

joshheald committed:
13175 Refactor POS image retrieval

There’s no need for us to use UIImage wrappers for a SwiftUI-only experience – this moves to using the `Image(_ name:)` initialiser directly

joshheald committed:
13175 Remove POS prefix from image name vars

The enum gives us the namespace information now

joshheald committed:
13175 Add whitespace into image name definitions

dangermattic commented:
<!--
  0 Errors
  2 Warnings: View files have been modified,...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/216">19.8</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

joshheald committed:
13175 fix whitespace

joshheald committed:
Fix whitespace

staskus commented on WooCommerce/WooCommerceTests/Extensions/IconsTests.swift:
```diff
@@ -966,37 +967,38 @@ final class IconsTests: XCTestCase {
 
     func test_posNewTransactionImage_is_not_nil() {
-        XCTAssertNotNil(UIImage.posNewTransactionImage)
+        XCTAssertNotNil(Image(PointOfSaleAssets.newTransactionImageName))
```
🤔 Writing a test for every added image is easy to forget.

To test this logic automatically, we could make `enum PointOfSaleAssets: CaseIterable`, define image names as cases, and have a single test that iterates through all the cases and checks if every one of them results in a non-nil image. What do you think?

joshheald replied:
Hmm... a little conflicted initially, but I like the idea.

The tests will obviously be a lot better, but I wanted to avoid having to write `.rawValue` when we use the asset name. This can be avoided a few ways though – I'll take a look.

I'll also separate the tests to a new file, since we will want to move them out if we ever make POS a separate app.

staskus replied:
It also may be an unpopular idea but we could also skip writing tests such as these. I don't remember a situation where an asset has been deleted by accident, I assume these tests defend from situations such as this.

joshheald replied:
Also against mistyped names of the assets, yeah.

I know what you mean... but if we can make it a lightweight test, that we don't need to rememeber, as you suggest, I think it'll have value. We don't get any help from the compiler on this so a test has more value than usual.

staskus reviewed (approved):
Thanks for these improvements and thanks for the convenient test cases!

Left one suggestion 👍

joshheald committed:
13175 Move POS asset tests to new file

joshheald committed:
13175 Use cases for POS assets

joshheald committed:
13175 Test for missing images using caseiterable

Using a loop in this test avoids us needing to remember to add each asset to the test.

Also, the previous tests in this branch could never fail 🫣

joshheald committed:
Merge branch 'trunk' into issue/13175-refactor-point-of-sale-image-retrieval

# Conflicts:
#	WooCommerce/WooCommerce.xcodeproj/project.pbxproj

staskus commented on WooCommerce/WooCommerceTests/POS/Tools/PointOfSaleAssetsTests.swift:
```diff
@@ -0,0 +1,17 @@
+        /// `Image(_ name:)` in production POS code. `Image` doesn't let us test whether the asset
+        /// exists or not, where `UIImage` will return nil if we can't find the asset in the bundle.
+        for asset in PointOfSaleAssets.allCases {
+            XCTAssertNotNil(
```
🎉
